src/bindings/python/PyGpuShaderDesc.cpp

```diff
@@ -1,186 +1,337 @@
+                                 { height * width * numChannels },
+                                 { sizeof(float) },
+                                 values);
+            },
+             "index"_a)
+
+        // 3D lut related methods
+        .def("getNum3DTextures", &GpuShaderDesc::getNum3DTextures)
+        .def("add3DTexture", [](GpuShaderDescRcPtr & self,
+                                const std::string & textureName, 
+                                const std::string & samplerName, 
+                                const std::string & uid,
+                                unsigned edgelen,
+                                Interpolation interpolation,
+                                const py::buffer & values)
+            {
+                py::buffer_info info = values.request();
+                checkBufferType(info, py::dtype("float32"));
+                checkBufferSize(info, edgelen*edgelen*edgelen*3);
+
+                py::gil_scoped_release release;
+
+                self->add3DTexture(textureName.c_str(), 
+                                   samplerName.c_str(), 
+                                   uid.c_str(),
+                                   edgelen,
+                                   interpolation,
+                                   static_cast<float *>(info.ptr));
+            },
+             "textureName"_a, "samplerName"_a, "uid"_a, "edgelen"_a, "interpolation"_a, "values"_a)
+        .def("get3DTextures", [](GpuShaderDescRcPtr & self) 
+            {
+                return Texture3DIterator(self);
+            })
+        .def("get3DTextureValues", [](GpuShaderDescRcPtr & self, unsigned index) 
+            {
+                py::gil_scoped_release release;
+
+                const char * textureName;
+                const char * samplerName;
+                const char * uid;
+                unsigned edgelen;
+                Interpolation interpolation;
+                self->get3DTexture(index, textureName, samplerName, uid, edgelen, interpolation);
+
+                const float * values;
+                self->get3DTextureValues(index, values);
+
+                py::gil_scoped_acquire acquire;
+                
+                return py::array(py::dtype("float32"), 
+                                 { edgelen*edgelen*edgelen * 3 },
+                                 { sizeof(float) },
+                                 values);
+            },
+             "index"_a);
+
+    py::class_<UniformIterator>(cls, "UniformIterator")
+        .def("__len__", [](UniformIterator & it) { return it.m_obj->getNumUniforms(); })
+        .def("__getitem__", [](UniformIterator & it, int i) 
+            { 
+                // GpuShaderDesc provides index check with exception
+                const char * name;
+                DynamicPropertyRcPtr value;
+                it.m_obj->getUniform(i, name, value);
+
+                return py::make_tuple(name, value);
```

Comment:
Is the 'name' pointer is correctly managed?

Reply:
The default `return_value_policy` for pointers is `take_ownsership`, so this should be safe. To be sure I just ran some tests where I returned the value from this iterator, then deleted the `GpuShaderDesc` and `UniformIterator`, and the tuple remained intact.